lib/string.js

```diff
@@ -295,5 +316,181 @@ internals.String.prototype.trim = function () {
     return obj;
 };
 
+function createUriRegex(optionalScheme) {
+    /**
+     * DIGIT = %x30-39 ; 0-9
+     */
+    var digit = '0-9',
+        digitOnly = '[' + digit + ']';
+
+    /**
+     * ALPHA = %x41-5A / %x61-7A   ; A-Z / a-z
+     */
+    var alpha = 'a-zA-Z',
+        alphaOnly = '[' + alpha + ']';
+
+    /**
+     * HEXDIG = DIGIT / "A" / "B" / "C" / "D" / "E" / "F"
+     */
+    var hexDigit = digit + 'A-Fa-f',
+        hexDigitOnly = '[' + hexDigit + ']';
+
+    /**
+     * unreserved = ALPHA / DIGIT / "-" / "." / "_" / "~"
+     */
+    var unreserved = alpha + digit + '-\\._~';
+
+    /**
+     * pct-encoded = "%" HEXDIG HEXDIG
+     */
+    var pctEncoded = '%' + hexDigit;
+
+    /**
+     * sub-delims = "!" / "$" / "&" / "'" / "(" / ")" / "*" / "+" / "," / ";" / "="
+     */
+    var subDelims = '!\\$&\'\\(\\)\\*\\+,;=';
+
+    /**
+     * pchar = unreserved / pct-encoded / sub-delims / ":" / "@"
+     */
+    var pchar = unreserved + pctEncoded + subDelims + ':@',
+        pcharOnly = '[' + pchar + ']';
+
+    /**
+     * elements separated by forward slash ("/") are alternatives.
+     */
+    var or = '|';
+
+    /**
+     * Rule to support zero-padded addresses.
+     */
+    var zeroPad = '0?';
+
+    /**
+     * dec-octet   = DIGIT                 ; 0-9
+     *            / %x31-39 DIGIT         ; 10-99
+     *            / "1" 2DIGIT            ; 100-199
+     *            / "2" %x30-34 DIGIT     ; 200-249
+     *            / "25" %x30-35          ; 250-255
+     */
+    var decOctect = '(' + zeroPad + zeroPad + digitOnly + or + zeroPad + '[1-9]' + digitOnly + or + '1' + digitOnly + digitOnly + or + '2' + '[0-4]' + digitOnly + or + '25' + '[0-5])';
+
+    /**
+     * scheme = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )
+     */
+    var scheme = alphaOnly + '[' + alpha + digit + '+-\\.]*';
+
+    // If we were passed one or more schemes, use them instead of the generic one.
+    if (optionalScheme) {
+        if (Array.isArray(optionalScheme)) {
+            optionalScheme = optionalScheme.join('|');
```

Comment:
Good catch! Added a test case around it and the code to handle it using `Hoek.escapeRegex()`